### ISSUE TITLE
Base64 images in URL Encoder

### ISSUE THREAD

**Issue by Veltinator** [thephpleague/commonmark#417]
It seems like that base64 images get also encoded with the unescapeAndEncode function. I noticed that this leads often to an exhaust of the memory_size in PHP. The preg_split in this functions is causing this. Is there any solution for this?

**Comment by Veltinator** [comment]
It seems like that base64 images get also encoded with the unescapeAndEncode function. I noticed that this leads often to an exhaust of the memory_size in PHP. The preg_split in this functions is causing this. Is there any solution for this?

**Comment by colinodell** [comment]
Based on some tests I did using PHP 7.4, this library seems to consume approximately 10 bytes of (peak) memory for each base64-encoded character, which doesn't seem too bad, compared to typical image sizes and PHP memory limits.  But it's certainly possible that my test case isn't doing something that yours is.

Would it be possible for you to send me some sample Markdown that you're experiencing this issue with?

**Comment by Veltinator** [comment]
I tried it with different sizes of images. With nearly FullHD and higher images it starts to exhaust my default memory_size of 128mb. I understand why to encode the URLs, but is it necessary to encode base64 too?

**Comment by colinodell** [comment]
> I understand why to encode the URLs, but is it necessary to encode base64 too?

It's because the library can't prevent people from passing in completely invalid base64 URLs like this:

```md
![](data:image/jpeg;charset=utf-8;base64,LookAtThisEmojiThatsNotEvenBase64☃️AndHere'sSome<html>)
```

Because it's possible that we might receive this as input, we have an obligation to double-check the URL and make sure it's normalized/encoded properly before rendering it in HTML.

That being said, proper base64-encoded strings _shouldn't_ contain special characters at all, so we can add a shortcut that checks whether the URL only contains non-special characters and return it as-is - something like this should work:

```php
public static function unescapeAndEncode(string $uri): string
{
    // Optimization: if the URL only includes characters we know will be kept as-is, then just return the URL as-is.
    if (\preg_match('/^[A-Za-z0-9~!@#$&*()\-_=+;:,.\/?]+$/', $uri)) {
        return $uri;
    }

    // ... the rest of the code goes here
}
```

I'll throw a PR together for this.

**Comment by colinodell** [comment]
Implemented in #418

**Comment by Veltinator** [comment]
Really cool how responsive and helpful you are! I appreciate it very much. I tested this and it works. Thank you!

**Comment by colinodell** [comment]
Awesome!!  I'll get this tagged and released as v1.3.2.

**Comment by colinodell** [comment]
Done! https://github.com/thephpleague/commonmark/releases/tag/1.3.2
